File: C++/soln-stack-problems/remove-duplicate-letters.cpp
/**
 @author Farheen Bano
  
 References-
 https://leetcode.com/problems/remove-duplicate-letters
 https://leetcode.com/problems/smallest-subsequence-of-distinct-characters
*/

class Solution {
public:
    string removeDuplicateLetters(string s) {
        unordered_map<char,int> map;
        for(char ch:s)
            map[ch]++;
        
        vector<bool> visited(26,false);
        stack<char> st;

        for(char ch:s){
            map[ch]--;
            
            if(visited[ch-'a']) 
                continue;
            
            while(!st.empty() && st.top()>ch && map[st.top()]>0){
                char removed=st.top();
                st.pop();    
                visited[removed-'a']=false;  
            }
            st.push(ch);
            visited[ch-'a']=true;
        }
        
        string ans="";
        while(!st.empty()){
            ans+=st.top();
            st.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: Python/soln-stack-problems/remove-duplicate-letters.py
'''
 @author Farheen Bano
  
 References-
 https://leetcode.com/problems/remove-duplicate-letters
 https://leetcode.com/problems/smallest-subsequence-of-distinct-characters
'''

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        s_map = dict()
        for ch in s:
            if ch in s_map:
                s_map[ch] += 1
            else:
                s_map[ch] = 1
   
        stack = []
        visited = set()
        
        for ch in s:
            if ch not in visited:
                while stack and stack[-1] > ch and s_map[stack[-1]] > 1:
                    s_map[stack[-1]] -= 1
                    visited.remove(stack.pop())
                    
                stack.append(ch)
                visited.add(ch)
            else:
                s_map[ch] -= 1
        
        return ''.join(stack)
        